src/hooks/UrqlUseSubscription.re

```diff
@@ -95,3 +95,63 @@ let useSubscription =
     (handler, args, parse),
   );
 };
+
+/**
+ * The functor implementation of useQuery. An alternative to the function API.
+ *
+ * Accepts the following arguments:
+ *
+ * Query - a graphql_ppx or graphql_ppx_re module containing the
+ * type t of the GraphQL query, the query string of the GraphQL query,
+ * and a parse function for decoding the JSON response.
+ */
+module type SubscriptionConfig = {
+  type t;
+  let query: string;
+  let parse: Js.Json.t => t;
+};
+
+module type MakeSubscriptionType = {
+  type resp;
+
+  let useSubscription:
+    (~variables: Js.Json.t=?, ~handler: handler('acc, resp, 'ret)) =>
+    UrqlTypes.hookResponse('ret);
+};
+
+module MakeSubscription =
+       (Subscription: SubscriptionConfig)
+       : (MakeSubscriptionType with type resp = Subscription.t) => {
```

Comment:
This one through me for a bit of a loop. I had to learn a bit about _sharing contraints_ in OCaml – see more in the **Destructive Substitution** section here: https://v1.realworldocaml.org/v1/en/html/functors.html Basically, we need to be able to pass the type `t` of the `module` passed _into_ the `MakeSubscription` functor to the `module type MakeSubscriptionType` such that the `resp` type is inferred properly. Quite fun!